### PR TITLE
[WIP] Add 'accepted' emoji for verified comments which are previously accepted answers

### LINKED ISSUE #5203: listing answered questions after deprecating Answers
Hi @ViditChitkara -- just want to open an issue to tackle how this will work:

https://publiclab.org/questions/answered/

Both on a) how we'd list them on this page, and also b) how answered comments are represented when viewing a list of comments in a question page?

I know you've thought this through, i just want to try to sum it up in one place as we take the next space. 

Thanks!

### PR DESCRIPTION
**Pull request by ViditChitkara**
Fixes #5203

### CODE REVIEW

**Comment by jywarren** [comment]
Possibly failing due to https://github.com/publiclab/plots2/pull/5256#issuecomment-476350660 -- sorry, fixing soon!

**Comment by jywarren** [comment]
This is great though - can you test it out on unstable, and then we can merge once we fix the tests? Thanks Vidit!!!

**Comment by ViditChitkara** [comment]
Sure, will push on unstable.

**Comment by jywarren** [comment]
Did this work out on unstable? Thanks!!!

**Comment by ViditChitkara** [comment]
Oh!! I missed out pushing it to unstable. Will do it right now.

**Comment by ViditChitkara** [comment]
@jywarren I pushed to unstable, I guess the build went well too!! But there is an error while visiting http://jenkins.laboratoriopublico.org/ and unstable has a quite older version of master. Any ideas why it might be happening?

**Comment by jywarren** [comment]
Ah, i believe @icarito fixed this -- can you try again? It was a Jenkins
issue.

**Comment by ViditChitkara** [comment]
Yeah, it's working now!! Trying again to push on unstable.

**Comment by ViditChitkara** [comment]
@jywarren I think the build went well. But I see no accepted emoji with comments. Would you mind taking a look at the migration and suggest if anything is wrong?

This is how it looks on localhost, though
Screenshot 2019-03-30 at 1 45 35 AM

File: db/migrate/20190325064225_create_accepted_comment_likes.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class CreateAcceptedCommentLikes < ActiveRecord::Migration[5.2]
+  def change
+    def up
+      answers = Answer.where(accepted: true)
+      timestamp_to_answer_id = answers.group('created_at').minimum('id')
+                                         .inject({}) {|hash, (key, value)| hash[key.to_i] = value; hash}
```
Comment (jywarren) [inline db/migrate/20190325064225_create_accepted_comment_likes.rb]: Oh, is this supposed to be `inject {|hash, (key, value)| ...` ? Sorry, a little hard to read...

File: db/migrate/20190325064225_create_accepted_comment_likes.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+      Comment.where(timestamp: timestamp_to_answer_id.keys).or(Comment.where(comment: answers.pluck(:content))).each do |c|
```
Comment (jywarren) [inline db/migrate/20190325064225_create_accepted_comment_likes.rb]: is `.or` ok with both being true? I assume so, maybe there's `.xor` for the other case...

File: db/migrate/20190325064225_create_accepted_comment_likes.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        c.likes.create(emoji_type: "Accepted")
```
Comment (jywarren) [inline db/migrate/20190325064225_create_accepted_comment_likes.rb]: Ah! Capitalization? Above this is `"accepted"`, no?

**Comment by jywarren** [comment]
Maybe i found something?

We can test on this page after re-running on unstable: https://unstable.publiclab.org/questions/msolomon/11-08-2018/only-seeing-part-of-color-spectrum

**Comment by jywarren** [comment]
Thanks @ViditChitkara !!! 👍

**Comment by ViditChitkara** [comment]
@jywarren  By any chance can we have something out the following?
1. ``` 
      answers = Answer.where(accepted: true)
      timestamp_to_answer_id = answers.group('created_at').minimum('id')
                                         .inject({}) {|hash, (key, value)| hash[key.to_i] = value; hash}

      Comment.where(timestamp: timestamp_to_answer_id.keys).or(Comment.where(comment: answers.pluck(:content))).count

Can we confirm if the last line gives any record (Count will do)?

2. Was there any error which occurred while running the migration
3. `Like.where(emoji_type: "Accepted").count` This will confirm if corresponding likes were created.

File: db/migrate/20190325064225_create_accepted_comment_likes.rb (hunk 1 above)
Comment (ViditChitkara) [reply db/migrate/20190325064225_create_accepted_comment_likes.rb]: yes, hash is the new injected hash and key, value gives corresponding keys and values at each iteration

File: db/migrate/20190325064225_create_accepted_comment_likes.rb (hunk 2 above)
Comment (ViditChitkara) [reply db/migrate/20190325064225_create_accepted_comment_likes.rb]: yes, so even if both are true, it'll still get the record.

File: db/migrate/20190325064225_create_accepted_comment_likes.rb (hunk 3 above)
Comment (ViditChitkara) [reply db/migrate/20190325064225_create_accepted_comment_likes.rb]: By the convention we are following it has to be Accepted, while comparison and all, it is converted to downcase so it isn't a problem.

**Comment by jywarren** [comment]
> Can we confirm if the last line gives any record (Count will do)?

yes, it returns 351 records on production!

I don't think there was an error - you can check https://jenkins.laboratoriopublico.org/job/Plots-Stable/ to see if it errored in stable? 

Hmm, i'm not sure how to run these on unstable... let me check...

**Comment by jywarren** [comment]
Sorry, I wasn't able to figure out how to run this in unstable, Vidit -- is there any way to simulate this locally? Or perhaps @icarito may be able to help us run this? I wasn't able to figure out what user to run it as -- i tried `jenkins` but it said gems weren't installed, with:

`warren@tycho:/srv/plots_unstable/plots2$ RAILS_ENV=production docker-compose -f containers/docker-compose-unstable.yml run web rails c`

and with `sudo su -l jenkins` and:

`jenkins@tycho:/srv/plots_unstable/plots2$ RAILS_ENV=production docker-compose -f containers/docker-compose-unstable.yml run web rails c`

☹️  sorry, this part of our containers work is not my forte!

**Comment by SidharthBansal** [comment]
There are some conflicts in your pr. It will be great if you can rebase your pr.
Thanks for the work done here at PL

**Comment by ViditChitkara** [comment]
@SidharthBansal will do!

**Comment by ViditChitkara** [comment]
@icarito any help on this? @jywarren do you think there may be a case that accepted answers didn't exist on unstable? Is there a way you could test this code on your local copy of public lab with some dummy data? Finding it hard to catch the bug here!!

**Comment by SidharthBansal** [comment]
There are some conflicts in your pr. It will be great if you can rebase your pr.
Thanks for the work done here at PL

**Comment by ViditChitkara** [comment]
@jywarren any ideas on this?

**Comment by jywarren** [comment]
Oh my apologies! Will look soon, sorry I'm tied up today. Could we ask
other reviewers for now? Thanks a lot Vidit!!!

**Comment by gauravano** [comment]
Hi @jywarren @ViditChitkara, I have restarted build here as system tests were failed but `rake test:all` was passed so let's see how it goes this time.

**Comment by ViditChitkara** [comment]
> Hi @jywarren @ViditChitkara, I have restarted build here as system tests were failed but rake test:all was passed so let's see how it goes this time.

@gauravano  the tests are passing, however, this wasn't running on unstable as we expected. Yet to figure out why was it happening.

**Comment by gauravano** [comment]
Oh, so test passed this time. That's some progress :sweat_smile:. So, what's primary road block here?

**Comment by ViditChitkara** [comment]
@gauravano the main problem we are facing is that the migration to add accepted emoji is not properly running on unstable or maybe the checkmark emojis are not appearing on comments(converted from answers). It works on local, though.

**Comment by cesswairimu** [comment]
Hi @ViditChitkara seems like you solved this on https://github.com/publiclab/plots2/pull/5478. Feel free to reopen this if its not. Thanks

**Pull request closed without merging**